engine/grid.py

```diff
@@ -301,79 +320,379 @@ def find_border_nodes(self):
+                            leftmost_node = node
+                            leftmost_node_pos = (row, col)
+
+            self.border_nodes = border_list
+            self.leftmost_node = leftmost_node
+            self.leftmost_node_pos = leftmost_node_pos
+
+ # --------------------- HELPER FUNCTIONS FOR TRAVERSAL CREATION -------------- #
+    def edge_column_of_next_row(self, pos):
+        """
+        If the direction we are heading is Left, returns the column of the innermost 
+        (closer to the center of the shape) leftmost node that is 
+            1) active  
+            2) not border node 
+        between the current row in [pos] and the next row above [pos].
+
+        If the direction we are heading is Right, returns the column of the innermost 
+        (closer to the center of the shape) rightmost node that is 
+            1) active  
+            2) not border node 
+        between the current row in [pos] and the next row above [pos].
+
+        If the direction we are heading is Up, returns the row of the innermost 
+        (closer to the center of the shape) top node that is 
+            1) active  
+            2) not border node 
+        between the current col in [pos] and the next col above [pos].
+
+        If the direction we are heading is Down, returns the row of the innermost 
+        (closer to the center of the shape) bottom node that is 
+            1) active  
+            2) not border node 
+        between the current col in [pos] and the next col above [pos].
+
+        If there are no more nodes in the next row, return None.
+
+        pos: the current position (col,row)
+        """
+        # Obtain all border nodes in the next and current rows 
+        if self.direction == self.Direction.RIGHT or self.direction == self.Direction.LEFT:
+            # Note: node_info: (node, row, col)
+            #Obtain all border_nodes on the next col
+            
+            candidate_nodes_next = [node_info for node_info in self.border_nodes if node_info[2] == pos[1] + 1]
+            candidate_nodes_curr = [node_info for node_info in self.border_nodes if node_info[2] == pos[1]]
+        else:
+            #Obtain all border_nodes on the next row
+            candidate_nodes_next = [node_info for node_info in self.border_nodes if node_info[1] == pos[0] + 1]
+            #all border_nodes on current row
+            candidate_nodes_curr = [node_info for node_info in self.border_nodes if node_info[1] == pos[0]]
+        
+        if (candidate_nodes_next == [] and candidate_nodes_curr == []):
+            # There are no border nodes in the current row or next row
             return None
+        elif (candidate_nodes_next == []):
+            # Since there we have reached the last row that we can traverse, 
+            # let's traverse this last row
+            if self.direction == self.Direction.RIGHT:
+                node_info_curr = max(candidate_nodes_curr, key=lambda node_info: node_info[1])
```

Comment:
what is node_info_curr?

Reply:
It is the last node we would traverse in the grid. Comments added